content/blog/jetstream-java-client-03-consume.md

```diff
@@ -0,0 +1,131 @@
++++
+categories = ["Clients", "Java", "JetStream"]
+date = "2021-05-24"
+tags = ["java", "stream", "consumer"]
+title = "JetStream Consumers with the NATS.io Java Library"
+author = "Scott Fauerbach"
++++
+
+The previous entry in this series showed us how to publish messages.
+This entry will start building up to subscribing by describing the configuration for consumers.
+
+## Push vs Pull Subscription
+
+Consumers, once created, can subscribe as either a `push` or `pull` mode. The server sends messages to push consumers while pull consumers have to ask for messages.
+This is mentioned here first because consumer configuration options may apply differently depending on the type of subscription.
+
+## Durable vs Ephemeral
+
+A consumer can be durable or ephemeral.
+
+A durable consumer is intended to be long-lived and the server will keep track of where the consumer is in the stream.
+So if the consumer stops then restarts, it will automatically restart where it left off and the configuration used to initialize
+the consumer will be remembered. A durable consumer is required when making a pull subscription and is optional
+when making a push subscription.
+
+An ephemeral consumer will only be tracked by the server while it is active and showing interest in the stream.
+Using consumer configuration options, an ephemeral can be manually set to start at a specific sequence or time.
+Ephemeral consumers are only allowed in push subscriptions.
+
+## Configuration Object
+The `ConsumerConfiguration` object has a builder to simplify setting options. Here is the builder skeleton showing all the builder methods.
+
+```
+ConsumerConfiguration c = ConsumerConfiguration.builder()
+    .durable(...)
+    .deliverPolicy(...)
+    .startSequence(...)
+    .startTime(...)
+    .deliverSubject(...)
+    .ackPolicy(...)
+    .ackWait(...)
+    .maxAckPending(...)
+    .replayPolicy(...)
+    .maxDeliver(...)
+    .filterSubject(...)
+    .rateLimit(...)
+    .sampleFrequency(...)
+    .idleHeartbeat(...)
+    .flowControl(...)
+    .build();
+```
+
+### Durable (Name)
+
+There is no default for the durable name. For push consumers, this is the only way to set the durable consumer name.
+For pull consumers, you can set the durable name in the `ConsumerConfiguration` object or in the `PullSubscribeOptions` object.
+
+### Deliver Policy / Start Sequence / Start Time
+
+When a consumer is first created, it can specify where in the stream it wants to start receiving messages.
+This is the `DeliverPolicy` and it's options are as follows:
+
+| Option  | Description  |
+| --- | --- |
+| All | All is the default policy. The consumer will start receiving from the earliest available message. |
+| Last | The consumer will start receiving messages with the last message. /
```

Comment:
I've found people dont get what "last message" means in this context, might need to expand a bit

Reply:
will do